File: chanb/DatabaseEngine.vb
Imports System.Data.SqlClient
Imports MySql.Data.MySqlClient
Imports System.Data.Common

Public Module DatabaseEngine

    Dim dbi As New DBInitializer

    Sub New()
        If Not (dbType = "mysql" Or dbType = "mssql") Then Throw New Exception(dbTypeInvalid)
    End Sub

    ''' <summary>
    ''' Execute text query.
    ''' </summary>
    ''' <param name="QueryText"></param>
    ''' <returns>Number of rows affected</returns>
    ''' <remarks></remarks>
    Public Function ExecuteNonQuery(ByVal QueryText As String) As Integer
        Select Case dbType
            Case "mssql"
                Dim i As New SqlConnection(dbi.ConnectionString)
                Dim command As New SqlCommand(QueryText, i)
                i.Open()
                Dim d As Integer = command.ExecuteNonQuery
                i.Close()
                Return d
            Case "mysql"
                Dim i As New MySqlConnection(dbi.ConnectionString)
                Dim command As New MySqlCommand(QueryText, i)
                i.Open()
                Dim d As Integer = command.ExecuteNonQuery
                i.Close()
                Return d
            Case Else
                Return Nothing
        End Select
    End Function

    ''' <summary>
    ''' Execute DbCommand query.
    ''' </summary>
    ''' <param name="Query"></param>
    ''' <returns>Number of rows affected</returns>
    ''' <remarks></remarks>
    Public Function ExecuteNonQuery(ByVal Query As DbCommand) As Integer
        Select Case dbType
            Case "mssql"
                Dim i As New SqlConnection(dbi.ConnectionString)
                i.Open()
                Query.Connection = i
                Dim d As Integer = Query.ExecuteNonQuery
                i.Close()
                Return d
            Case "mysql"
                Dim i As New MySqlConnection(dbi.ConnectionString)
                i.Open()
                Query.Connection = i
                Query.Prepare()
                Dim d As Integer = Query.ExecuteNonQuery
                i.Close()
                Return d
            Case Else
                Return Nothing
        End Select
    End Function

    ''' <summary>
    ''' Execute text query with an existing connection.
    ''' </summary>
    ''' <param name="QueryText"></param>
    ''' <returns>Number of rows affected</returns>
    ''' <remarks></remarks>
    Public Function ExecuteNonQuery(ByVal QueryText As String, ByVal i As DbConnection) As Integer
        Select Case dbType
            Case "mssql"
                Dim command As New SqlCommand(QueryText, CType(i, SqlConnection))
                Dim d As Integer = command.ExecuteNonQuery
                Return d
            Case "mysql"
                Dim command As New MySqlCommand(QueryText, CType(i, MySqlConnection))
                Dim d As Integer = command.ExecuteNonQuery
                Return d
            Case Else
                Return Nothing
        End Select
    End Function

    ''' <summary>
    ''' Execute DbCommand query with an existing connection.
    ''' </summary>
    ''' <param name="Query"></param>
    ''' <returns>Number of rows affected</returns>
    ''' <remarks></remarks>
    Public Function ExecuteNonQuery(ByVal Query As DbCommand, ByVal i As DbConnection) As Integer
        Select Case dbType
            Case "mssql"   
                Query.Connection = i
                Dim d As Integer = Query.ExecuteNonQuery
                Return d
            Case "mysql"
                Query.Connection = i
                Dim d As Integer = Query.ExecuteNonQuery
                Return d
            Case Else
                Return Nothing
        End Select
    End Function

    ''' <summary>
    ''' Execute text query.
    ''' </summary>
    ''' <param name="QueryText"></param>
    ''' <returns>ChanbQuery</returns>
    ''' <remarks></remarks>
    Public Function ExecuteQueryReader(ByVal QueryText As String) As ChanbQuery
        Select Case dbType
            Case "mssql"
                Dim i As New SqlConnection(dbi.ConnectionString)
                Dim command As New SqlCommand(QueryText, i)
                i.Open()
                Dim reader As SqlDataReader = command.ExecuteReader
                Dim c As New ChanbQuery
                c.Connection = i
                c.Reader = reader
                Return c
            Case "mysql"
                Dim i As New MySqlConnection(dbi.ConnectionString)
                Dim command As New MySqlCommand(QueryText, i)
                i.Open()
                Dim reader As MySqlDataReader = command.ExecuteReader
                Dim c As New ChanbQuery
                c.Connection = i
                c.Reader = reader
                Return c
            Case Else
                Return Nothing
        End Select
    End Function

    ''' <summary>
    ''' Execute DbCommand query.
    ''' </summary>
    ''' <param name="Query"></param>
    ''' <returns>ChanbQuery</returns>
    ''' <remarks></remarks>
    Public Function ExecuteQueryReader(ByVal Query As DbCommand) As ChanbQuery
        Select Case dbType
            Case "mssql"
                Dim i As New SqlConnection(dbi.ConnectionString)
                Query.Connection = i
                i.Open()
                Dim reader As SqlDataReader = CType(Query.ExecuteReader, SqlDataReader)
                Dim c As New ChanbQuery
                c.Connection = i
                c.Reader = reader
                Return c
            Case "mysql"
                Dim i As New MySqlConnection(dbi.ConnectionString)
                Query.Connection = i
                i.Open()
                Dim reader As MySqlDataReader = CType(Query.ExecuteReader, MySqlDataReader)
                Dim c As New ChanbQuery
                c.Connection = i
                c.Reader = reader
                Return c
            Case Else
                Return Nothing
        End Select
    End Function

    ''' <summary>
    ''' Execute text query with an existing connection.
    ''' </summary>
    ''' <param name="QueryText"></param>
    ''' <param name="i"></param>
    ''' <returns>ChanbQuery</returns>
    ''' <remarks></remarks>
    Public Function ExecuteQueryReader(ByVal QueryText As String, ByVal i As DbConnection) As ChanbQuery
        Select Case dbType
            Case "mssql"
                Dim command As New SqlCommand(QueryText, CType(i, SqlConnection))
                Dim reader As SqlDataReader = command.ExecuteReader
                Dim c As New ChanbQuery
                c.Connection = i
                c.Reader = reader
                Return c
            Case "mysql"
                Dim command As New MySqlCommand(QueryText, CType(i, MySqlConnection))
                Dim reader As MySqlDataReader = command.ExecuteReader
                Dim c As New ChanbQuery
                c.Connection = i
                c.Reader = reader
                Return c
            Case Else
                Return Nothing
        End Select
    End Function

    ''' <summary>
    ''' Execute DbCommand query with an existing connection.
    ''' </summary>
    ''' <param name="Query"></param>
    ''' <param name="i"></param>
    ''' <returns>ChanbQuery</returns>
    ''' <remarks></remarks>
    Public Function ExecuteQueryReader(ByVal Query As DbCommand, ByVal i As DbConnection) As ChanbQuery
        Select Case dbType
            Case "mssql"
                Query.Connection = i
                Dim reader As SqlDataReader = CType(Query.ExecuteReader, SqlDataReader)
                Dim c As New ChanbQuery
                c.Connection = i
                c.Reader = reader
                Return c
            Case "mysql"
                Query.Connection = i
                Dim reader As MySqlDataReader = CType(Query.ExecuteReader, MySqlDataReader)
                Dim c As New ChanbQuery
                c.Connection = i
                c.Reader = reader
                Return c
            Case Else          
                Return Nothing
        End Select
    End Function

    Public Function GenerateDbCommand() As DbCommand
        Select Case dbType
            Case "mssql"
                Dim dbConnection As New SqlConnection(dbi.ConnectionString)
                Dim dbC As New SqlCommand()
                dbConnection.Open()
                dbC.Connection = dbConnection
                Return dbC
            Case "mysql"
                Dim dbConnection As New MySqlConnection(dbi.ConnectionString)
                Dim dbC As New MySqlCommand()
                dbConnection.Open()
                dbC.Connection = dbConnection
                Return dbC
            Case Else
                Return Nothing
        End Select
    End Function

    Public Function MakeParameter(ByVal name As String, ByVal value As Object, ByVal type As DbType) As DbParameter
        Select Case dbType
            Case "mssql"
                Dim p As New SqlParameter(name, type)
                p.Value = value
                Return p
            Case "mysql"
                Dim p As New MySqlParameter(name, value)
                Select Case type
                    Case Data.DbType.String
                        p.MySqlDbType = MySqlDbType.Text
                    Case Data.DbType.DateTime
                        p.MySqlDbType = MySqlDbType.DateTime
                    Case Data.DbType.Int32
                        p.MySqlDbType = MySqlDbType.Int32
                    Case Data.DbType.AnsiString
                        p.MySqlDbType = MySqlDbType.Text
                End Select
                Return p
            Case Else
               Return Nothing
        End Select
    End Function


End Module

Public Class ChanbQuery

    Public Reader As IDataReader
    Public Connection As DbConnection

End Class

File: src/chanb/Database/DatabaseEngine.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Data.Common;
using System.Data.SqlClient;
using MySql.Data.MySqlClient;

using chanb.Settings;

namespace chanb.Database
{
    public static class DatabaseEngine
    {
        //public static int ExecuteNonQuery(string queryText)
        //{
        //    switch (DatabaseSettings.DbType)
        //    {
        //        case Enums.DatabaseType.MsSQL:
        //            using (SqlConnection sq = new SqlConnection(DatabaseSettings.ConnectionString))
        //            {
        //                sq.Open();
        //                using (SqlCommand sc = new SqlCommand(queryText, sq))
        //                {
        //                    return sc.ExecuteNonQuery();
        //                }
        //            }
        //        case Enums.DatabaseType.MySQL:
        //            using (MySqlConnection sq = new MySqlConnection(DatabaseSettings.ConnectionString))
        //            {
        //                sq.Open();
        //                using (MySqlCommand sc = new MySqlCommand(queryText, sq))
        //                {
        //                    return sc.ExecuteNonQuery();
        //                }
        //            }
        //        default:
        //            return 0;
        //    }
        //}

        public static int ExecuteNonQuery(DbCommand query, DbConnection dc)
        {
            switch (DatabaseSettings.DbType)
            {
                case Enums.DatabaseType.MsSQL:
                case Enums.DatabaseType.MySQL:
                    using (query)
                    {
                        query.Connection = dc;
                        return query.ExecuteNonQuery();
                    }
                default:
                    return -1;
            }
        }
   
        public static DbCommand GenerateDbCommand()
        {
            switch (DatabaseSettings.DbType)
            {
                case Enums.DatabaseType.MsSQL:
                    SqlConnection sc = new SqlConnection(DatabaseSettings.ConnectionString);
                    sc.Open();
                    return new SqlCommand() { Connection = sc };
                case Enums.DatabaseType.MySQL:
                    MySqlConnection my_sc = new MySqlConnection(DatabaseSettings.ConnectionString);
                    my_sc.Open();
                    return new MySqlCommand() { Connection = my_sc };

                default:
                    return null;
            }
        }

        public static DbCommand GenerateDbCommand(string commandText)
        {
            DbCommand d = GenerateDbCommand();
            d.CommandText = commandText;
            return d;
        }

        public static DbCommand GenerateDbCommand(DbConnection conx)
        {
            DbCommand d = GenerateDbCommand();
            d.Connection = conx;
            return d;
        }

        public static DbCommand GenerateDbCommand(string commandText, DbConnection conx) 
        {
            DbCommand dc = GenerateDbCommand(commandText);
            dc.Connection = conx;
            return dc;
        }

        public static DbParameter MakeParameter(string Name, object value, System.Data.DbType Type)
        {
            switch (DatabaseSettings.DbType)
            {
                case Enums.DatabaseType.MsSQL:
                    return new SqlParameter(Name, Type) { Value = value };
                case Enums.DatabaseType.MySQL:
                    MySqlParameter p = new MySqlParameter(Name, value);
                    switch (Type) 
                    {
                        case System.Data.DbType.String:
                            p.MySqlDbType = MySqlDbType.Text;
                            break;
                        case System.Data.DbType.DateTime:
                            p.MySqlDbType = MySqlDbType.DateTime;
                            break;
                        case System.Data.DbType.Int32:
                            p.MySqlDbType = MySqlDbType.Int32;
                            break;
                        case System.Data.DbType.AnsiString:
                            p.MySqlDbType = MySqlDbType.Text;
                            break;
                        case System.Data.DbType.Boolean:
                            p.MySqlDbType = MySqlDbType.Bit;
                            break;
                        case System.Data.DbType.Int64:
                            p.MySqlDbType = MySqlDbType.Int64;
                            break;
                        default:
                            break;
                    }
                    return p;
                default:
                    return null;
            }
        }

        public static DbConnection GetDBConnection() 
        {
            switch (DatabaseSettings.DbType) 
            {
                case Enums.DatabaseType.MsSQL:
                    return new SqlConnection(DatabaseSettings.ConnectionString);
                case Enums.DatabaseType.MySQL:
                    return new MySqlConnection(DatabaseSettings.ConnectionString);
                default:
                    return null;
            }
        }

    }
}